microservices-connector/internal/controller/gmconnector_controller.go

```diff
@@ -377,18 +415,102 @@ func (r *GMConnectorReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	//to start a router service
 	//in case the graph changes, we need to apply the changes to router service
 	//so we need to apply the router config every time
+	totalService += 1
 	err := reconcileRouterService(ctx, r.Client, graph)
 	if err != nil {
 		return reconcile.Result{Requeue: true}, errors.Wrapf(err, "Failed to reconcile router service")
+	} else {
+		successService += 1
 	}
 
 	graph.Status.Status = fmt.Sprintf("%d/%d/%d", successService, externalService, totalService)
-	if err = r.Status().Update(context.TODO(), graph); err != nil {
+	if err = r.Status().Update(ctx, graph); err != nil {
 		return reconcile.Result{Requeue: true}, errors.Wrapf(err, "Failed to Update CR status to %s", graph.Status.Status)
 	}
+
+	if updateExistGraph {
+		//check if the old annotations are still in the new graph
+		for k := range oldAnnotations {
+			if _, ok := graph.Status.Annotations[k]; !ok {
+				//if not, remove the resource from k8s
+				r.deleteRecordedResource(k, ctx)
+			}
+		}
+	}
+
 	return ctrl.Result{}, nil
 }
 
+func (r *GMConnectorReconciler) deleteRecordedResource(key string, ctx context.Context) {
+	kind := strings.Split(key, ":")[0]
+	apiVersion := strings.Split(key, ":")[1]
+	name := strings.Split(key, ":")[2]
+	ns := strings.Split(key, ":")[3]
+	obj := &unstructured.Unstructured{}
+	obj.SetKind(kind)
+	obj.SetName(name)
+	obj.SetNamespace(ns)
+	obj.SetAPIVersion(apiVersion)
+	err := r.Delete(ctx, obj)
+	// the resource may have been deleted by other means, i.e. user manually delete or delete namespace
+	// ignore the error if delete failed i.e resource not found
+	// since I don't want to block the process for not clearing the finalizer
+	if err != nil {
+		fmt.Printf("Failed to delete resource %s:%s:%s: %v\n", ns, kind, name, err)
+	} else {
+		fmt.Printf("Success to delete %s:%s:%s\n", ns, kind, name)
+	}
+}
+
+func recordResourceStatus(graph *mcv1alpha3.GMConnector, step *mcv1alpha3.Step, obj *unstructured.Unstructured) (uint, error) {
+	// var statusStr string
+	var success uint = 0
+	// graph.SetFinalizers(append(graph.GetFinalizers(), fmt.Sprintf("%s-.-%s-.-%s", obj.GetKind(), obj.GetNamespace(), obj.GetName())))
+	// save the resource name into annotation for status update and resource management
+	graph.Status.Annotations[fmt.Sprintf("%s:%s:%s:%s", obj.GetKind(), obj.GetAPIVersion(), obj.GetName(), obj.GetNamespace())] = "provisioned"
+
+	if obj.GetKind() == Service {
```

Comment:
Hi @KfreeZ, is it more simply if we can separately handle `Service` and `Deployment`?